Creating library and clients for querying jenkins.

mohammedzee1000 opened the pull request:
- Create utils for commonly used functions.
- Create base Client structure.
- Create base ApiClient Structure.
- Create basic OpenshiftClient Structure.
- Create basic Jenkins API Client structure.

Signed-off-by: Mohammed Zeeshan Ahmed <mohammed.zee1000@gmail.com>

mohammedzee1000 committed:
Creating library and clients for querying jenkins.

- Create utils for commonly used functions.
- Create base Client structure.
- Create base ApiClient Structure.
- Create basic OpenshiftClient Structure.
- Create basic Jenkins API Client structure.

Signed-off-by: Mohammed Zeeshan Ahmed <mohammed.zee1000@gmail.com>

mohammedzee1000 committed:
Moving url conversion of job list to allow parented jobs.

mohammedzee1000 committed:
Adding functions to query important endpoints of jenkins wf api.

mohammedzee1000 committed:
Adding docs.

mohammedzee1000 committed:
Adding more docs and renaming functions.

mohammedzee1000 committed:
Renaming ApiClient to APIClient.

navidshaikh commented on ccp/lib/utils.py:
```diff
@@ -0,0 +1,82 @@
+
+    try:
+        response = requests.get(
+            request, params=params, verify=verify_ssl, headers=headers
```
add `auth`

mohammedzee1000 replied:
Added, thanks for pointing it out

mohammedzee1000 committed:
Adding auth to request_url.

mohammedzee1000 committed:
Setup jenkins token command to use common core oc cmd.

navidshaikh commented on ccp/lib/clients/base.py:
```diff
@@ -0,0 +1,58 @@
+            secure=True,
+            verify_ssl=True,
+            authorization=None,
+            server="localhost",
```
I would say, let's not use default for `server` and keep it a mandatory field for initialization, every client shall have a server to contact to.

Same applies to port. Keep the server and port values mandatory.

The good targets for optional fields are for eg: [secure, verify_ssl, authorizatio, etc]

mohammedzee1000 replied:
Updating for server, port can be optional as URL can be without the port, in case of default ports for service or port hidden behind ```oc expose```.

navidshaikh commented on ccp/lib/clients/base.py:
```diff
@@ -0,0 +1,58 @@
+    ):
+        self.server_endpoint = "{}://{}{}".format(
+            "https" if secure else "http",
+            server if server else "localhost",
```
this `server` conditional need to change as per above comment and update.

mohammedzee1000 replied:
I think this check should be there as user can still explicitly pass ```None```.

navidshaikh replied:
IMO, if the user passes None, the code should identify that, report and exit.

Doing it for `localhost` is an assumption, which is not good.
We can instead say that in help/docstring that, possible values are ["localhost", "0.0.0.0", "127.0.0.1"]

navidshaikh commented on ccp/lib/clients/base.py:
```diff
@@ -0,0 +1,58 @@
+            server if server else "localhost",
+            ":{}".format(
+                port
+            ) if port else ""
```
if you are planning to keep `port` optional, the default value you used i.e. ("none") will fail the logic here.

if you are planning to keep `port` optional, set its default value to `None`

mohammedzee1000 replied:
Done, yes we want to keep this optional, reason stated above.

navidshaikh commented on ccp/lib/clients/base.py:
```diff
@@ -0,0 +1,58 @@
+        """
+        Given a target (query without url), queries endpoint and gets data
+        :param target: The query to send to the api server, including after
+        destination
```
`including after detination` ? re-word please.

mohammedzee1000 replied:
reworded

navidshaikh commented on ccp/lib/clients/base.py:
```diff
@@ -0,0 +1,58 @@
+        )
+        headers = {}
+
+        self.authorization.add_header(headers)
```
why headers are empty ?

navidshaikh replied:
say if someone initialized with authorization=True, we override it with empty headers then ?

mohammedzee1000 replied:
Sorted, now ```self.authorization``` will be used only if its of type ```ccp.lib.clients.utils.authorization.Authorization```

navidshaikh commented on ccp/lib/clients/jenkins/base.py:
```diff
@@ -0,0 +1,43 @@
+
+    def __init__(
+            self,
+            server="localhost",
```
again, no defaults for server value for client side implementation (which shall require a server to talk to)

mohammedzee1000 replied:
Addressed

navidshaikh commented on ccp/lib/clients/jenkins/base.py:
```diff
@@ -0,0 +1,43 @@
+            server=server,
+            port=port,
+            authorization=BearerAuthorization(
+                OpenshiftCmdClient().get_jenkins_service_account_token()
```
3 level of object creation in same operation, I would recommend, slit the operations in multiple operations or at least it needs comments.

mohammedzee1000 replied:
Separated out. FYI i just did not want to create unnecessary variables. My policy, if a something is not being used at least twice in a place, it does not need a variable :)

navidshaikh commented on ccp/lib/clients/jenkins/base.py:
```diff
@@ -0,0 +1,43 @@
+        """
+        dest = ""
+        for i in jobs_ordered_list:
+            dest = "{}/job/{}".format(
```
using the value in loop which itself is being processed ?

I did not understand the purpose of this method ?

mohammedzee1000 replied:
Well the job can be nested. This function builds the nested job endpoint from list of job names. I know name can be misleading, maybe i will change that.

navidshaikh commented on ccp/lib/clients/jenkins/client.py:
```diff
@@ -0,0 +1,91 @@
+"""
+This file contains ready to use Jenkins Clients.
+"""
```
add a line spacing

mohammedzee1000 replied:
Added

navidshaikh commented on ccp/lib/clients/jenkins/client.py:
```diff
@@ -0,0 +1,91 @@
+
+class OpenshiftJenkinsWorkflowAPIClient(OpenshiftJenkinsBaseAPIClient):
+    """
+    Helps query the jenkins workflow api endpoints.
```
replace all api with `API`

mohammedzee1000 replied:
Done

navidshaikh commented on ccp/lib/clients/jenkins/client.py:
```diff
@@ -0,0 +1,91 @@
+
+    def __init__(
+            self,
+            server="localhost",
```
no defaults for server param please

mohammedzee1000 replied:
Sorted

navidshaikh commented on ccp/lib/clients/jenkins/client.py:
```diff
@@ -0,0 +1,91 @@
+    def __init__(
+            self,
+            server="localhost",
+            port=None
```
this is the right way to initialize a param if its processed later by `if` conditional

mohammedzee1000 replied:
Yes

navidshaikh commented on ccp/lib/clients/jenkins/client.py:
```diff
@@ -0,0 +1,91 @@
+
+    def describe_build_run(
+            self, ordered_jobs_list, build_number
+    ):
```
this could have been done in single line ? :-)

navidshaikh replied:
```
In [1]: len("    def describe_build_run(self, ordered_jobs_list, build_number):")
Out[1]: 66
```

as long as length is less than 80, keep it in same line

mohammedzee1000 replied:
Sorted

navidshaikh commented on ccp/lib/clients/jenkins/client.py:
```diff
@@ -0,0 +1,91 @@
+
+    def get_logs_of_execution_node(
+            self, ordered_jobs_list, build_number, node_number
+    ):
```
docstring missing

mohammedzee1000 replied:
Added

navidshaikh commented on ccp/lib/clients/openshift/client.py:
```diff
@@ -0,0 +1,43 @@
+    """
+
+    def __init__(self):
+        self.core_oc_cmd = "oc"
```
specify the complete path of `oc`,

mohammedzee1000 replied:
Done

navidshaikh commented on ccp/lib/clients/openshift/client.py:
```diff
@@ -0,0 +1,43 @@
+
+    def get_jenkins_service_account_token(self):
+        """
+        Queries and gets the jenkins token. This assumes you have access to oc
```
*Jenkins

mohammedzee1000 replied:
Done

navidshaikh commented on ccp/lib/clients/openshift/client.py:
```diff
@@ -0,0 +1,43 @@
+        """
+        Queries and gets the jenkins token. This assumes you have access to oc
+        command line.
+        :return: The token, if it was able to get it.
```
else return what ?

navidshaikh replied:
add that description as well

navidshaikh commented on ccp/lib/clients/openshift/client.py:
```diff
@@ -0,0 +1,43 @@
+        token_cmd = """
+        {core_oc_cmd} sa/jenkins --template='{template1}' | xargs -n 1 
+        {core_oc_cmd} get secret --template='{template2}' | head -n 1 | 
+        base64 -d -
```
you don't want new lines in the command ? use `\`

navidshaikh commented on ccp/lib/clients/openshift/client.py:
```diff
@@ -0,0 +1,43 @@
+        :return: The token, if it was able to get it.
+        """
+        token = None
+        token_cmd = """
```
also skip the extra space added before command, use `\`

mohammedzee1000 replied:
Done

navidshaikh commented on ccp/lib/clients/openshift/client.py:
```diff
@@ -0,0 +1,43 @@
+        """.format(
+            core_oc_cmd=self.core_oc_cmd,
+            template1="{{range .secrets}}{{ .name }} {{end}}",
+            template2="{{ if .data.token }}{{ .data.token }}{{end}}'"
```
try rendering this string in ipython?

ensure the double/single/desired quotes are preserved after rendering the string using `format` and as needed by `--template`

mohammedzee1000 replied:
We should be good
```
In [9]: token_cmd

In [10]:         token_cmd = """{core_oc_cmd} sa/jenkins --template='{template1}' | \
    ...:         xargs -n 1 {core_oc_cmd} get secret --template='{template2}' | \
    ...:         head -n 1 | base64 -d -\
    ...:         """.format(
    ...:             core_oc_cmd="/usr/bin/oc",
    ...:             template1="{{range .secrets}}{{ .name }} {{end}}",
    ...:             template2="{{ if .data.token }}{{ .data.token }}{{end}}"
    ...:         )

In [11]: token_cmd
Out[11]: "/usr/bin/oc sa/jenkins --template='{{range .secrets}}{{ .name }} {{end}}' | xargs -n 1 /usr/bin/oc get secret --template='{{ if .data.token }}{{ .data.token }}{{end}}' | head -n 1 | base64 -d -"

In [12]: 

```

navidshaikh commented on ccp/lib/clients/openshift/client.py:
```diff
@@ -0,0 +1,43 @@
+        else:
+            data, err = out
+            if data:
+                token = data
```
`data.strip()`

navidshaikh replied:
and you can just return here itself

mohammedzee1000 replied:
Done

navidshaikh commented on ccp/lib/clients/openshift/client.py:
```diff
@@ -0,0 +1,43 @@
+                token = data
+            else:
+                raise Exception("Failed to get token due to {}".format(err))
+        return token
```
no need to return at the end

mohammedzee1000 replied:
Removed

navidshaikh commented on ccp/lib/clients/utils/authorization.py:
```diff
@@ -0,0 +1,19 @@
+        headers["Authorization"] = self.authorization_header_string
+
+
+class BearerAuthorization(Authorization):
```
docstring

mohammedzee1000 replied:
Added

navidshaikh commented on ccp/lib/utils.py:
```diff
@@ -0,0 +1,83 @@
+        request, params=None, verify_ssl=False, auth=None, headers=None
+):
+    """
+    Queries a specified url and returns data, if any
```
URL

mohammedzee1000 replied:
sorted

navidshaikh commented on ccp/lib/utils.py:
```diff
@@ -0,0 +1,83 @@
+    :param request: The url to send the request to.
+    :type request str
+    :param params: Any params that are to be passed on to the request
+    :param verify_ssl: IF true, ssl errors are ignored
```
`s/IF/If`

and `If False, ssl errors are ignored` ?

mohammedzee1000 replied:
Sorted

navidshaikh commented on ccp/lib/utils.py:
```diff
@@ -0,0 +1,83 @@
+    Queries a specified url and returns data, if any
+    :param request: The url to send the request to.
+    :type request str
+    :param params: Any params that are to be passed on to the request
```
`type params` is missing

mohammedzee1000 replied:
Sorted

mohammedzee1000 commented:
Due to the nature of this code, unit tests may not be possible as it requires a jenkins and openshift server setup. It requires functional tests, which requires functional tests to be setup on pipeline.

navidshaikh commented on ccp/lib/utils.py:
```diff
@@ -0,0 +1,83 @@
+    :param headers: Any extra headers that you wish to pass along.
+    :type headers dict
+    :return: The response object, or None upon failure, alongwith exception
+    message, if any or None
```
return should return either object or raise exception(which includes the message in it, handler can utilize), not the mixture of raise+message

navidshaikh replied:
consumer will check if its a response or object

mohammedzee1000 replied:
Ok, exception is now raised so should not be a problem.

navidshaikh commented on ccp/lib/utils.py:
```diff
@@ -0,0 +1,83 @@
+        )
+    except Exception as ex:
+        response = None
+        exception = ex
```
I would suggest, instead of return exception, raise the exception and let the consumer handle it.

mohammedzee1000 replied:
Done

navidshaikh commented on ccp/lib/utils.py:
```diff
@@ -0,0 +1,83 @@
+    return response, exception
+
+
+def encode(data):
```
just follow the good practice you followed throughout the PR and add docstring

mohammedzee1000 replied:
Done

navidshaikh commented on ccp/lib/utils.py:
```diff
@@ -0,0 +1,83 @@
+    return base64.b64encode(data)
+
+
+def decode(data):
```
just follow the good practice you followed throughout the PR and add docstring

mohammedzee1000 replied:
Added

navidshaikh commented on ccp/lib/utils.py:
```diff
@@ -0,0 +1,83 @@
+def run_cmd(cmd, shell=False, use_pipes=True):
+    """
+    Runs the command that is passed to it via subprocess.Popen
+    :param cmd: The command that needs to be executed
```
add `type param` whether needed string or list

navidshaikh commented on ccp/lib/utils.py:
```diff
@@ -0,0 +1,83 @@
+    exception = None
+    try:
+        p = subprocess.Popen(
+            cmd, shell=shell, stdout=subprocess.PIPE, stderr=subprocess.PIPE
```
keep using pipe default as False. If the JSON response output is big (which could be case for build info), the pipe size increases and it could block the calls

navidshaikh commented on ccp/lib/utils.py:
```diff
@@ -0,0 +1,83 @@
+    """
+    Parses the json and loads the data
+    :param data:
+    :return:
```
missing description

mohammedzee1000 replied:
Added

mohammedzee1000 committed:
Fixing navids comments on clients implementation.

mohammedzee1000 committed:
Updating templates to explicitly mention stages for parallel.

The explicit naming gives us option to correctly filter
data via the workflow API.
As a side affect, this helps align the templates to
recommended approach

Signed-off-by: Mohammed Zeeshan Ahmed <mohammed.zee1000@gmail.com>

mohammedzee1000 commented:
Needs https://github.com/dharmit/ccp-openshift/pull/43

mohammedzee1000 committed:
Fixing command to get oc token.

mohammedzee1000 committed:
Fixing oc command for token.

mohammedzee1000 committed:
Adding utility function to use ast parsing.

mohammedzee1000 committed:
Fixing json parse function.

mohammedzee1000 committed:
Cleaned up code.
